### DISCUSSION TITLE
cannot create queries using "contains"

### DISCUSSION THREAD

**Discussion by spemmons** [gravitational/teleport#24180] [Q&A] [answered]
i have been able to successfully use all query functions -- equals, exists, search -- except "contains" ... the documentation provides no examples but mentions the use of "fields" and "values" as arguments ... i cannot seem to find any "fields" that are accepted ... everything i try generates an error

i was hoping that i could use "contains" to find a partial string match for labels associated w/ my servers, but maybe that's not how it works ... no use of `labels[...]` or `hostname` seem to work ... not sure what a valid "field" would be

any suggestions of working uses of "contains" that i can try?

**Comment by scottpgallagher** [comment]
I believe we have it documented here: https://goteleport.com/docs/reference/predicate-language/#resource-filtering

**Reply by spemmons** [reply]
as i mentioned, the documentation linked above provides no examples for "contains" while it does for _every other_ query function ... and the _2 sentences_ of documentation mentions "fields" while providing no definition _anywhere_ for what a valid "field" would be ... as mentioned, i tried using `labels` which works w/ other functions like "exists" and `hostname` which is displayed as a column name and subject to the "search" function, but i cannot find _anything_ that works w/ "contains"

**Comment by webvictim** [comment, marked as answer]
You can find the test cases for the predicate language here: https://github.com/gravitational/teleport/blob/master/lib/services/parser_test.go/#L30-L157

All the examples in those tests seem to use `session.particpants` - as in `contains(session.participants, "username")`

I believe `participants` comes from here, and this may give a better list of some other "fields" you can use with `contains`: https://github.com/gravitational/teleport/blob/master/lib/services/parser.go/#L425-L449
